### PR TITLE
🌱 Use GitHub forms for issue templates

### LINKED ISSUE #7815: Move all existing issue templates to GitHub forms
**User Story**
There is a possibility to move all existing templates we have in the repo from markdown format to [GitHub forms for issue templates ](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/syntax-for-issue-forms)which is basically YAML representation of the template. Example issue template in kubernetes repo to get the idea of how it looks like: https://github.com/kubernetes/kubernetes/issues/new?assignees=&labels=kind%2Ffailing-test&template=failing-test.yaml 
kubernetes repo has already switched to the GitHub forms some time ago. See more discussion on https://github.com/kubernetes-sigs/cluster-api/pull/7778#issuecomment-1367431498

/kind feature

### PR DESCRIPTION
**Pull request by furkatgofurov7**
You can see the results here: https://github.com/furkatgofurov7/cluster-api/issues/new/choose

**What this PR does / why we need it**:
Moves markdown-based issue templates (bug_report, failing-test, feature_request, and flaking-test) to YAML format using [GitHub forms for issue templates ](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/syntax-for-issue-forms)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7815

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+labels: kind/bug
+body:
+  - type: textarea
+    id: problem
+    attributes:
+      label: What steps did you take and what happened?
+      description: |
+        Please provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner.
+        If this matter is security related, please follow the guidelines described in https://github.com/kubernetes-sigs/cluster-api/blob/main/SECURITY_CONTACTS
+      placeholder: "[A clear and concise description on how to REPRODUCE the bug.]"
```
Comment (fabriziopandini) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:12]: q: can we drop brackets around the placeholder texts?

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      label: What did you expect to happen?
+    validations:
+      required: true
+
+  - type: textarea
+    id: repro
+    attributes:
+      label: How can we reproduce it (as minimally and precisely as possible)?
+    validations:
+      required: false
```
Comment (fabriziopandini) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:28]: nit
```suggestion
```
since the majority of not required fields doesn't have required explicitly set to false

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      placeholder: "[Miscellaneous information that will assist in solving the issue.]"
+
+  - type: textarea
+    id: capiVersion
+    attributes:
+      label: Cluster-api version
+      value: |
+        <details>
+
+        </details>
```
Comment (fabriziopandini) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:43]: Q: what is the scope/benefit of using  details here?
Also, ideally, this should be a combo with the list of releases, but I'm not sure we can get this automatically filled (I don't wat to add to the release tasks)

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      label: Kubernetes version
+      value: |
+        <details>
+
+        ```console
+        $ kubectl version
+        # paste output here
+        ```
+
+        </details>
```
Comment (fabriziopandini) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:59]: Q: is there a way to provide instructions in a way that the user is not forced to clean them up when adding a value? what about using placeholder?

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 5]
```diff
@@ -0,0 +1,91 @@
+name: 🐛 Bug Report
+description: Report a bug encountered while using Cluster API
+labels: kind/bug
```
Comment (fabriziopandini) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:3]: Q: sometimes uses files issues as a kind/bug but then we switch to something else, like e.g kind/support or kind/documentation etc. Will this be possible also when using forms?

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 6]
```diff
@@ -1,29 +0,0 @@
...
-
-**Environment:**
-
-- Cluster-api version:
-- minikube/kind version:
-- Kubernetes version: (use `kubectl version`):
-- OS (e.g. from `/etc/os-release`):
-
-/kind bug
-[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]
```
Comment (fabriziopandini) [inline .github/ISSUE_TEMPLATE/bug_report.md:29]: Is there a way to allow users to add area labels in the issue?

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        <details>
+
+        </details>
+    validations:
+      required: true
+
+  - type: textarea
+    id: kubeVersion
+    attributes:
+      label: Kubernetes version
```
Comment (fabriziopandini) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:50]: wondering if we really need this in the form...
Also, sometimes Kubernetes version is hard to interpret in a CAPI context. is it the Kubernetes version of the kind cluster,  Kubernetes version of the management cluster version or Kubernetes version a workload cluster?
I'm kind of leaning toward keeping the form simple... opinions? 
(same applies to Minikube/Kind version and to osVersion)

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    id: repro
+    attributes:
+      label: How can we reproduce it (as minimally and precisely as possible)?
+    validations:
+      required: false
+
+  - type: textarea
+    id: additional
+    attributes:
+      label: Anything else you would like to add?
```
Comment (fabriziopandini) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:33]: Q: Should we move this to the last position in the form?

**Review by fabriziopandini — COMMENTED** [review]
Great to see this happening.
Is there a way to see a preview of the forms?

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 3 above)
Comment (sbueringer) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:43]: I think having a simple field for Cluster API version where folks can enter the version is good enough

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 5 above)
Comment (sbueringer) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:3]: Not when we use the labels field here. If we add it as one of the text fields it should be preserved 1:1

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 7 above)
Comment (sbueringer) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:50]: Not sure but keeping it simple seems fine. In most issues those fields are not really filled out although we could enforce it here now (but then it might not make sense in all cases...)

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 3 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:43]: > Q: what is the scope/benefit of using details here? Also, ideally, this should be a combo with the list of releases, but I'm not sure we can get this automatically filled (I don't wat to add to the release tasks)

This I have copy/pasted from the upstream k8s issue template. Not sure about automatic fill up but there is an option to have a drop down where we could specify list of releases

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 4 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:59]: Sure, I will try to use a placeholder if possible or simply remove the instructions since it is obvious that output should be under the command :)

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 5 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:3]: Good question, I have not thought about it. I will try that use case out and come back, as @sbueringer text field should be the way forward

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 6 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.md:29]: Would preserving the same text at the end of the form work?

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 7 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:50]: +1 keeping it simple as well, though these 3 (k8s, kind/minikube/os versions) were in the current template as well so I kept them. Sometimes it might be useful to know at least k8s version at least? I think we should make them as optional and leave to the user to fill it out or leave.

**Comment by fabriziopandini** [comment]
note: there is also the option to introduce forms incrementally starting from the less controversial/the ones where we are more confident about the result...

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 1 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml]: Done

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 2 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml]: Dropped it, thanks.

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 3 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml]: Removed the "<details>"  and made it a simple text field

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 4 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml]: I moved the "$ kubectl version" to the placeholder

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 5 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml]: I have moved `labels: kind/bug` to the text field only in the bug_report issue template so it is preserved 1:1 as @sbueringer suggested

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 6 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.md:29]: @fabriziopandini area labels will be shown the same way as we had in soon be removed bug_report.md.

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 7 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:38]: I have kept the k8s version and made it optional, so if user reporting wants to report it, they can, otherwise left empty. Please let me know if we need to remove it completely, since I should admit there are different contexts for the k8s version

**Comment by furkatgofurov7** [comment]
> note: there is also the option to introduce forms incrementally starting from the less controversial/the ones where we are more confident about the result...

@fabriziopandini got it, let me try again to squeeze everything with the same changes, otherwise I will break down :)
Also, https://github.com/furkatgofurov7/cluster-api/issues/new?assignees=&labels=&template=bug_report.yaml is the example bug report (since this is the one most comments I had) in my fork which should be the exact copy of what we would have with these changes.

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 8 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml]: Done

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+  - type: textarea
+    id: repro
+    attributes:
+      label: How can we reproduce it (as minimally and precisely as possible)?
+
+  - type: textarea
+    id: capiVersion
+    attributes:
+      label: Cluster-api version
```
Comment (vincepri) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:30]:
```suggestion
      label: Cluster API Version
```

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    id: expected
+    attributes:
+      label: What did you expect to happen?
+    validations:
+      required: true
+
+  - type: textarea
+    id: repro
+    attributes:
+      label: How can we reproduce it (as minimally and precisely as possible)?
```
Comment (vincepri) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:25]: This part is technically asked in the first question

File: .github/ISSUE_TEMPLATE/feature_request.yaml [hunk 11]
```diff
@@ -0,0 +1,28 @@
+name: ✨ Feature request
+description: Suggest an idea for this project.
+labels: kind/feature
+body:
+  - type: textarea
+    id: user_story
+    attributes:
+      label: What would you like to be added (User Story)?
```
Comment (vincepri) [inline .github/ISSUE_TEMPLATE/feature_request.yaml:8]:
```suggestion
      label: What would you like to be added (User Stories, Outcomes, Goals)?
```

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 10 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:25]: @vincepri do you mean right at the beginning? if that is the case, we would start asking about the repro steps before letting the reporter define a problem they are facing & what they expected. Technically should be fine, but it might make more sense to keep the current order as is.

Comment (sbueringer) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:25]: I think it's just that this feels the same as 
```
placeholder: "A clear and concise description on how to REPRODUCE the bug."
```

I guess the question is, is "What steps did you take and what happened?" the same as the steps needed to reproduce. It sounds very similar to me, but no strong opinion.

If we think it's the same I would drop the "repro" field

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    attributes:
+      label: Anything else you would like to add?
+      placeholder: "Miscellaneous information that will assist in solving the issue."
+
+  - type: textarea
+    id: templateLabel
+    attributes:
+      label: Label(s) to be applied
+      value: |
+        kind/bug
```
Comment (sbueringer) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:52]:
```suggestion
        /kind bug
```
Otherwise Prow won't pick it up

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+      label: Kubernetes version
+      placeholder: "$kubectl version"
+
+  - type: textarea
+    id: additional
+    attributes:
+      label: Anything else you would like to add?
+      placeholder: "Miscellaneous information that will assist in solving the issue."
+
+  - type: textarea
```
Comment (sbueringer) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:47]: Wondering if we want something similar for the other issues as well

Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml:47]: Yes, done

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 12 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml]: Fixed, thank you!

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 10 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml]: @sbueringer @vincepri removed the repro field from the bug report template as suggested: https://github.com/kubernetes-sigs/cluster-api/pull/7819/commits/ec3f533f4c1533d933aebd73f46caae02977889f

File: .github/ISSUE_TEMPLATE/bug_report.yaml [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    id: expected
+    attributes:
+      label: What did you expect to happen?
+    validations:
+      required: true
+
+  - type: textarea
+    id: capiVersion
+    attributes:
+      label: Cluster-api version
```
Comment (sbueringer) [inline .github/ISSUE_TEMPLATE/bug_report.yaml:25]:
```suggestion
      label: Cluster API version
```
Given the previous finding from Vince. Did you maybe forgot to push some changes / lose some changes?
(please also adjust Cluster API in other places)

**Review by sbueringer — COMMENTED** [review]
Thx!

Just one nit otherwise lgtm from my side

File: .github/ISSUE_TEMPLATE/bug_report.yaml (hunk 14 above)
Comment (furkatgofurov7) [reply .github/ISSUE_TEMPLATE/bug_report.yaml]: Thanks for catching, looks like I messed up the previous commit, fixed it now.

**Comment by furkatgofurov7** [comment]
@sbueringer @vincepri @fabriziopandini thank you all for the reviews, this should be good to go, PTAL

**Comment by sbueringer** [comment]
Thx!

/lgtm

**Comment by sbueringer** [comment]
Let's get this merged, we can always iterate

/approve

**Comment by furkatgofurov7** [comment]
> Let's get this merged, we can always iterate
> 
> /approve

@sbueringer thank you! 

All seems to be working as expected by first look: https://github.com/kubernetes-sigs/cluster-api/issues/8265, please let me know if anything has to be changed @sbueringer @fabriziopandini @vincepri
